### PR TITLE
fix: fix failing tests

### PR DESCRIPTION
**Pull request by szevzol**
My first intention was to use your library in my production config handling, to take advantage of classes in the code when reading config data. (Thank you for your work!)

My issue was that I had "null" values in my yaml file and dataconf cannot handle them today as far as I've seen. However I removed those null values and decided to see how I can improve dataconf to handle null.

To be able to do that first I would like to fix failing unit tests.

I'm working on Win 10, using python 3.9.6.

### CODE REVIEW

File: tests/test_multi.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ class N:
         class A:
             a: N
 
-        assert multi.string("a { b = 1\nc = 2 }").string("a { c = 3\nd = 4 }").on(
+        assert multi.string("a { b = 1,\nc = 2 }").string("a { c = 3,\nd = 4 }").on(
```
Comment (zifeo) [inline tests/test_multi.py:39]: comma should not be required in the spec, likely the issue is with windows using `\r\n` and `\n`, can you try with that?

File: tests/test_regression.py [hunk 2]
```diff
@@ -142,9 +142,9 @@ def test_url(self) -> None:
         class A:
             url: str
 
-        os.environ["p_url"] = "https://github.com/zifeo/dataconf"
-        assert dataconf.env("p", A) == A(url="https://github.com/zifeo/dataconf")
-        os.environ.pop("p_url")
+        os.environ["P_URL"] = "https://github.com/zifeo/dataconf"
+        assert dataconf.env("P", A) == A(url="https://github.com/zifeo/dataconf")
```
Comment (zifeo) [inline tests/test_regression.py:146]: curious to hear if anything is breaking here?

File: tests/test_regression.py [hunk 3]
```diff
@@ -171,9 +171,12 @@ class Config:
 
         original = Config("test_dump")
 
-        with tempfile.NamedTemporaryFile() as f:
-            dataconf.dump(f.name, original, out="yaml")
-            validate = dataconf.file(f.name, Config)
+        temporary_file = tempfile.NamedTemporaryFile()
+        temporary_file_name = temporary_file.name
+        temporary_file.close()
```
Comment (zifeo) [inline tests/test_regression.py:176]: another windows specific behaviour, have a look [there](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) and see whether `delete=False` is enough? do not forget to clean the file

**Review by zifeo — CHANGES_REQUESTED** [review]

File: tests/test_multi.py (hunk 1 above)
Comment (szevzol) [reply tests/test_multi.py:39]: Even if I try with \r\n, the result is the same:

```
FAILED tests/test_multi.py::TestMulti::test_complex - pyparsing.exceptions.ParseSyntaxException: , found '='  (at char 13), (line:2, col:3)
```

Comment (zifeo) [reply tests/test_multi.py:39]: Make sure you did not upgrade vyparsing by mistake: https://github.com/zifeo/dataconf/blob/main/poetry.lock#L449

Comment (szevzol) [reply tests/test_multi.py:39]: I have pyparsing 3.1.1.

I'm going to debug it in details hopefully tonight.

File: tests/test_regression.py (hunk 2 above)
Comment (szevzol) [reply tests/test_regression.py:146]: In Windows, os module intentionally sets all keys to uppercase, and this is the reason why in Windows test case fails.

File: tests/test_regression.py (hunk 3 above)
Comment (szevzol) [reply tests/test_regression.py]: delete=False solved it

Comment (zifeo) [reply tests/test_regression.py]: Can you add a manual remove after the test has run? like a try/finally

Comment (szevzol) [reply tests/test_regression.py]: I've added a fixture to provide temp file and a teardown which will close the file

File: tests/test_multi.py (hunk 1 above)
Comment (szevzol) [reply tests/test_multi.py:39]: Very strange. This fails:

`        assert multi.string("a { b = 1\nc = 2 }").string("a { c = 3\nd = 4 }").on(
            A
        ) == A(a=N(b=1, c=3, d=4))`

But this, basically the same assertion, but passes:

`        assert multi.string("a { b = 1\nc = 2 }").string("a { d = 4\nc = 3 }").on(
            A
        ) == A(a=N(b=1, c=3, d=4))`

Difference is that in the second string c and d swapped.

Comment (zifeo) [reply tests/test_multi.py:39]: there are a lot of weirdness in the latest pyhocon change (that bumped pyparsing): https://github.com/chimpler/pyhocon/issues?q=is%3Aissue+is%3Aopen+0.3.60, I would stick at pyparsing v2 and those issues should be gone

Comment (szevzol) [reply tests/test_multi.py:39]: I can confirm that my tests are passing with pyhocon==0.3.59 and pyparsing==2.4.7. Should dataconf depend on these versions instead?

Comment (zifeo) [reply tests/test_multi.py:39]: Yes please and we can merge your PR 👍

Comment (szevzol) [reply tests/test_multi.py:39]: Fine! is it enough to update version in pyproject.toml?

Comment (szevzol) [reply tests/test_multi.py]: Updated also poetry.lock

**Review by zifeo — APPROVED** [review]

**Comment by zifeo** [comment]
@szevzol thanks for the contribution and happy 2024

**Pull request merged**